_posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

```diff
@@ -0,0 +1,182 @@
+
+In the end, both Document 1 and Document 2 are returned, and their scores are calculated based on the distances between the query vector and the vectors collected in the search queue.
+
+<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-after-1.png" alt="multi-vector-behavior-after" class="img-centered"/>
+
+Now, even with a k value set to 2, the search will return 2 documents, regardless of whether the 2 nearest vectors belong to one or multiple documents.  Additionally, this enhancement improves recall for users, as each segment now returns the k nearest documents instead of just the k nearest vectors of documents.
+## How to use nested field to store multi-vector
+To utilize nested fields for storing multi-vectors, ensure that your OpenSearch cluster has the KNN plugins installed. Now, let's delve into the process of creating a KNN index with nested fields and conducting searches on it.
+
+Begin by creating a KNN index, setting the knn value to true in the index settings. Additionally, define the knn_vector type within the nested type. All other parameters for the knn_vector type remain the same as those for a normal knn_vector type.
+```
+PUT my-knn-index
+{
+  "settings": {
+    "index": {
+      "knn": true
+    }
+  },
+  "mappings": {
+    "properties": {
+      "my_vectors": {
+        "type": "nested",
+        "properties": {
+          "my_vector": {
+            "type": "knn_vector",
+            "dimension": 2,
+            "method": {
+              "name": "hnsw",
+              "space_type": "l2",
+              "engine": "faiss"
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Next, insert your data. The number of data entries within a nested field is not fixed; you can index a different number of nested field items for each document. For instance, in this example, three nested field items were added for document 1, and two for document 2.
+```
+PUT _bulk?refresh=true
+{ "index": { "_index": "my-knn-index", "_id": "1" } }
+{"my_vectors":[{"my_vector":[1,1]},{"my_vector":[2,2]},{"my_vector":[3,3]}]}
+{ "index": { "_index": "my-knn-index", "_id": "2" } }
+{"my_vectors":[{"my_vector":[10,10]},{"my_vector":[20,20]}]}
+```
+
+When ready to search the data, note that the query structure differs slightly from a normal KNN search. Your query should be wrapped by a nested query with the specified path. Moreover, your field name should encompass both the nested field name and the knn_vector field name, concatenated with a dot.
+```
+GET my-knn-index/_search
+{
+  "query": {
+    "nested": {
+      "path": "my_vectors",
+      "query": {
+        "knn": {
+          "my_vectors.my_vector": {
+            "vector": [1,1],
+            "k": 2
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+With a k value of 2, you'll retrieve two documents instead of one, even if the three nearest vectors belong to document 1.
```

Comment:
```suggestion
When specifying a k value of 2, you'll retrieve two documents instead of one, even if the three nearest vectors belong to document 1:

```

Reply:
Is it `:`? not `.` at the end?